internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+// the job object will be terminated.
+func (job *jobObject) close() error {
+	closeErr := false
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			closeErr = true
+		}
+		job.jobHandle = 0
+	}
+	if job.iocpHandle != 0 {
+		if err := windows.Close(job.iocpHandle); err != nil {
+			closeErr = true
+		}
+		job.iocpHandle = 0
+	}
+	if closeErr {
+		return errors.New("failed to close one or more handles")
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *jobObject) assign(p *JobProcess) error {
+	if p.Pid() == 0 {
+		return errors.New("process has not started")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, uint32(p.Pid()))
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *jobObject) terminate() error {
+	if job.jobHandle != 0 {
+		return windows.TerminateJobObject(job.jobHandle, 1)
+	}
+	return nil
+}
+
+func (job *jobObject) shutdown(ctx context.Context) error {
+	pids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	var (
+		terminate bool
+		signalErr bool
+	)
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Error("failed to send ctrl-break to process in job")
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again. If there is any left then terminate the job.
+	newPids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.terminate()
+	}
+	return nil
+}
+
+// Returns all of the process IDs in the job object.
+func (job *jobObject) pids() ([]uint32, error) {
+	info := winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST{}
+	var retLen uint32
+	err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		&retLen,
+	)
+
+	// If there are no processes in the job the next call to QueryInformation will just
+	// hang until a memory alloc error. Return here.
+	if err == nil {
+		// Return empty slice instead of nil to play well with the caller of this.
+		// I don't think we should return an error on there being no PIDs.
+		var pids []uint32
+		return pids, nil
+	}
+
+	if err != nil && err != winapi.ERROR_MORE_DATA {
+		return nil, fmt.Errorf("failed initial query for PIDs in job object: %s", err)
+	}
+
+	buf := make([]uintptr, unsafe.Sizeof(info)+unsafe.Sizeof(info.ProcessIdList[0])*uintptr(info.NumberOfAssignedProcesses-1))
```

Comment:
I'm having a hard time understanding this line. 

My first concern is why this is made into a uintptr slice. My understanding is that you're trying to create a buffer that's meant to represent a JOBOBJECT_BASIC_PROCESS_ID_LIST struct and this buffer needs to have the appropriate amount of memory for internal use. Since `unsafe.Sizeof` returns the size in bytes of the object, I think this needs to be a byte slice instead. 

My second concern is if we can make the calculation cleaner. Since the use of `unsafe` is just for getting the size of an object, I think this calculation should just be moved into a separate variable. 

So something like this maybe? 
```suggestion
        jobBasicProcessIDListSize := unsafe.Sizeof(info)+(unsafe.Sizeof(info.ProcessIdList[0])*uintptr(info.NumberOfAssignedProcesses-1))
	buf := make([]byte, jobBasicProcessIDListSize)
```